src/transaction/log_replication_atomic.cpp

```diff
@@ -132,27 +150,17 @@ namespace cublog
 	  case LOG_START_ATOMIC_REPL:
 	    // nested atomic replication are not allowed
 	    assert (!m_atomic_helper.is_part_of_atomic_replication (header.trid));
-	    m_atomic_helper.add_atomic_replication_sequence (header.trid, m_redo_lsa, m_redo_context);
+	    m_atomic_helper.append_control_log (&thread_entry, header.trid, header.type, m_redo_lsa, m_redo_context);
 	    set_lowest_unapplied_lsa ();
 	    break;
 	  case LOG_END_ATOMIC_REPL:
 	    assert (m_atomic_helper.is_part_of_atomic_replication (header.trid));
-	    // non-sysop and sysop atomic replication sequences cannot mix
-	    assert (!m_atomic_helper.can_end_sysop_sequence (header.trid));
-	    m_atomic_helper.apply_and_unfix_atomic_replication_sequence (&thread_entry, header.trid);
+	    m_atomic_helper.append_control_log (&thread_entry, header.trid, header.type, m_redo_lsa, m_redo_context);
 	    set_lowest_unapplied_lsa ();
 	    break;
 	  case LOG_SYSOP_ATOMIC_START:
-	    if (m_atomic_helper.is_postpone_sequence_started (header.trid))
-	      {
-		// atomic sequence is part of a postpone sequence
-		// which is itself part of another, previosly started, atomic sequence
-	      }
-	    else
-	      {
-		m_atomic_helper.start_sysop_sequence (header.trid, m_redo_lsa, m_redo_context);
-		set_lowest_unapplied_lsa ();
-	      }
+	    m_atomic_helper.append_control_log (&thread_entry, header.trid, header.type, m_redo_lsa, m_redo_context);
```

Comment:
I think it needs assertion like LOG_START_ATOMIC_REPL. 
`assert (!m_atomic_helper.is_part_of_atomic_replication (header.trid));` 

LOG_SYSOP_ATOMIC_START is only logged once within a transaction before it is LOG_SYSOP_END with commit.

`LOG_SYSOP_ATOMIC_START - LOG_SYSOP_ATOMIC_START - LOG_SYSOP_END - LOG_SYSOP_END`
above scenario is not allowed, only the first LOG_SYSOP_ATOMIC_START can be logged. 

I'm still not clear on the relationship between LOG_SYSOP_ATOMIC_START and LOG_START_ATOMIC_REPL . Is there a case where two are nested? 
such as
`LOG_SYSOP_ATOMIC_START -> LOG_START_ATOMIC_REPL` or
`LOG_START_ATOMIC_REPL -> LOG_SYSOP_ATOMIC_START` ?

Reply:
> I think it needs assertion like LOG_START_ATOMIC_REPL.
assert (!m_atomic_helper.is_part_of_atomic_replication (header.trid));

> LOG_SYSOP_ATOMIC_START is only logged once within a transaction before it is LOG_SYSOP_END with commit.

See section 4. in the comment before `class atomic_replication_helper`.

> LOG_SYSOP_ATOMIC_START - LOG_SYSOP_ATOMIC_START - LOG_SYSOP_END - LOG_SYSOP_END
above scenario is not allowed, only the first LOG_SYSOP_ATOMIC_START can be logged.

I know. This is the ugliest case encountered yet.
It is handled in the `can_purge` function. See:
```
	else if (LOG_END_ATOMIC_REPL == last_entry.m_rectype)
	  {
	    // search backwards, until a start atomic replication log record is met; other 'sysop
	    // ends' are allowed and skipped;
	    // these are the sysops without a matching sysop atomic start - it is the consequence of the
	    // known fact (optimization?): when adding consecutive 'sysop atomic start' log records only
	    // one such log record is added
```

> I'm still not clear on the relationship between LOG_SYSOP_ATOMIC_START and LOG_START_ATOMIC_REPL . Is there a case where two are nested?
> such as
> LOG_SYSOP_ATOMIC_START -> LOG_START_ATOMIC_REPL or
> LOG_START_ATOMIC_REPL -> LOG_SYSOP_ATOMIC_START ?

Yes, there are at least 2 that I encountered with tests:
- one is documented in section 4 of the comment before `class atomic_replication_helper`
- the second one is more recent and has been added with extra atomic sequences added for `RVHF_SET_PREV_VERSION_LSA`